Objects/typeobject.c

```diff
@@ -402,33 +403,90 @@ _PyType_Name(PyTypeObject *type)
     return s;
 }
 
+
 static PyObject *
-type_name(PyTypeObject *type, void *context)
+_PyType_QualName(PyTypeObject *type)
 {
     if (type->tp_flags & Py_TPFLAGS_HEAPTYPE) {
         PyHeapTypeObject* et = (PyHeapTypeObject*)type;
-
-        Py_INCREF(et->ht_name);
-        return et->ht_name;
+        Py_INCREF(et->ht_qualname);
+        return et->ht_qualname;
     }
     else {
         return PyUnicode_FromString(_PyType_Name(type));
     }
 }
 
+
+static PyObject*
+_PyType_Module(PyTypeObject *type)
+{
+    PyObject *mod;
+
+    if (type->tp_flags & Py_TPFLAGS_HEAPTYPE) {
+        mod = _PyDict_GetItemId(type->tp_dict, &PyId___module__);
+        if (mod == NULL) {
+            PyErr_Format(PyExc_AttributeError, "__module__");
+            return NULL;
+        }
+        Py_INCREF(mod);
+    }
+    else {
+        const char *s = strrchr(type->tp_name, '.');
+        if (s != NULL) {
+            mod = PyUnicode_FromStringAndSize(
+                type->tp_name, (Py_ssize_t)(s - type->tp_name));
+            if (mod != NULL)
+                PyUnicode_InternInPlace(&mod);
+        }
+        else {
+            mod = _PyUnicode_FromId(&PyId_builtins);
+            Py_XINCREF(mod);
+        }
+    }
+    return mod;
+}
+
+
+PyObject *
+_PyType_FullName(PyTypeObject *type)
```

Comment:
In case of non-heap types you can just return `PyUnicode_FromString(type->tp_name)`.

In case of heap types the code can be a tiny bit simpler if inline `_PyType_Module()` and `_PyType_QualName()` because you can get rid of increfs/decrefs and NULL checks.

Reply:
> In case of non-heap types you can just return PyUnicode_FromString(type->tp_name).

Done. This change removes "builtins." when formatting builtins type. So %T of a string becomes "str" (instead of "builtins.str"). IMHO it's the expected behaviour.

> In case of heap types the code can be a tiny bit simpler if inline _PyType_Module() and _PyType_QualName() because you can get rid of increfs/decrefs and NULL checks.

I don't think that performance matters here. I prefer to reuse the same code, to make sure that type.__qualname__, type.__module__ and %T behave the same for heap types.